esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -5885,4 +5897,207 @@ int CWsDfuEx::GetIndexData(IEspContext &context, bool bSchemaOnly, const char* i
+
+        unsigned expirySecs = req.getExpirySeconds();
+
+        // JCSMORE - make this configurable
+        if (expirySecs > 86400)
+            expirySecs = 86400; //24 hours
+
+        // setup "expiryTime"
+        time_t simple;
+        time(&simple);
+        simple += expirySecs;
+        CDateTime expiryDt;
+        expiryDt.set(simple);
+        StringBuffer expiryTimeStr;
+        expiryDt.getString(expiryTimeStr);
+        Owned<IPropertyTree> secureMetaInfo = createPTreeFromIPT(&metaInfo);
+        secureMetaInfo->setProp("expiryTime", expiryTimeStr);
+
+        extractFilePartInfo(*secureMetaInfo, *fDesc);
+
+        // create random AES key and IV
+        char randomAesKey[aesKeySize];
+        char randomIV[aesBlockSize];
+        fillRandomData(aesKeySize, randomAesKey);
+        fillRandomData(aesBlockSize, randomIV);
+
+        // 1st serialize to JSON
+        StringBuffer secureMetaInfoJson;
+        toJSON(secureMetaInfo, secureMetaInfoJson);
+
+        // 2nd compress
+        MemoryBuffer compressedSecureMetaInfoMb;
+        fastLZCompressToBuffer(compressedSecureMetaInfoMb, secureMetaInfoJson.length(), secureMetaInfoJson.str());
+
+        // 3rd encrypt with AES key
+        MemoryBuffer encryptedSecureMetaInfoMb;
+        aesKeyEncrypt(encryptedSecureMetaInfoMb, compressedSecureMetaInfoMb.length(), compressedSecureMetaInfoMb.bytes(), randomAesKey, randomIV);
+
+        metaInfo.setPropBin("secureInfo", encryptedSecureMetaInfoMb.length(), encryptedSecureMetaInfoMb.bytes());
+
+        // 4th encrypt AES key with public *DAFILESRV* key
+        VStringBuffer publicKeyFName("%s.pub.pem", keyPairName); // JCSMORE - could cache loaded keys
+        Owned<CLoadedKey> publicKey = loadPublicKeyFromFile(publicKeyFName, nullptr);
+        MemoryBuffer encryptedAesKeyMb;
+        publicKeyEncrypt(encryptedAesKeyMb, aesKeySize, randomAesKey, *publicKey);
+
+        /* perhaps should combine into single prob..
+         * Does IV need to be generated/sent each time?
+         */
+        metaInfo.setPropBin("securityKey", encryptedAesKeyMb.length(), encryptedAesKeyMb.bytes());
+        metaInfo.setPropBin("securityIV", aesBlockSize, randomIV);
+    }
+    else
+#endif
+        extractFilePartInfo(metaInfo, *fDesc);
+}
+
+const char *getFileDafilesrvKeyName(IDistributedFile &file)
+{
+    /* JCSMORE - key needs to be looked up..
+     * potentially different per target group, but must be same for all dafilesrv's related to a file.
+     * [ unless we change scheme so that there is a separate encrypted blob per part - then we could have a different key per dafilesrv. ]
+     */
+    const char *keyPairName = "/home/jsmith/.ssh/id_rsa";
+
+    // JCSMORE - option to have authorization disabled in configuration
+    //keyPairName = nullptr;
+
+    return keyPairName;
+}
+
+void CWsDfuEx::getReadAccess(IEspContext &context, IUserDescriptor *udesc, IEspDFUReadAccessRequest &req, IEspDFUReadAccessResponse &resp)
```

Comment:
@wangkx - change to use read and populate interfaces for request/response directly - that made the implementation of this method a lot simpler/easier to read.

You mentioned this being shared with the 'create and publish' method.. let's discuss the commonality and minimizing duplication there.

Reply:
@jakesmith 2 items in DFUReadAccessRequest should not be shared. In the 'create and publish' method, The JobId has been removed from the request. The ReturnFileInfo is always true for the 'create and publish' method. I may create an EspStruct which contains other 6 items. The DFUReadAccessRequest may contains that EspStruct  + JobId + ReturnFileInfo . I may create another EspStruct which contains 7 items of the DFUReadAccessResponse. Should I go head to try that?

Reply:
Please check the last commit in https://github.com/wangkx/HPCC-Platform/tree/h20340.